frontend/src/metabase/visualizations/visualizations/Funnel.jsx

```diff
@@ -110,36 +111,49 @@ export default class Funnel extends Component {
       showColumnSetting: true,
       marginBottom: "0.625rem",
     }),
+    "funnel.order_dimension": {
+      getValue: (_series, settings) => settings["funnel.dimension"],
+      readDependencies: ["funnel.rows"],
+    },
     "funnel.rows": {
       section: t`Data`,
       widget: ChartSettingOrderedSimple,
-      isValid: (series, settings) => {
-        const funnelRows = settings["funnel.rows"];
 
-        if (!funnelRows || !_.isArray(funnelRows)) {
-          return false;
-        }
-        if (!funnelRows.every(setting => setting.originalIndex !== undefined)) {
-          return false;
+      getValue: (series, settings) => {
+        const seriesOrder = settings["funnel.rows"];
+        const seriesKeys = series.map(s => keyForSingleSeries(s));
+        const orderDimension = settings["funnel.order_dimension"];
+        const dimension = settings["funnel.dimension"];
+
+        const getDefault = keys =>
+          keys.map(key => ({
+            key,
+            name: key,
+            enabled: true,
+          }));
+        if (
+          !seriesOrder ||
+          !_.isArray(seriesOrder) ||
+          !seriesOrder.every(setting => setting.key !== undefined) ||
+          orderDimension !== dimension
+        ) {
+          return getDefault(seriesKeys);
         }
 
-        return (
-          funnelRows.every(setting => series[setting.originalIndex]) &&
-          funnelRows.length === series.length
-        );
-      },
+        const removeMissingOrder = (keys, order) =>
+          order.filter(o => keys.includes(o.key));
+        const newKeys = (keys, order) =>
+          keys.filter(key => !order.find(o => o.key === key));
 
-      getDefault: transformedSeries => {
-        return transformedSeries.map(s => ({
-          name: s.card.name,
-          originalIndex: s.card.originalIndex,
-          enabled: true,
-        }));
+        return [
+          ...removeMissingOrder(seriesKeys, seriesOrder),
+          ...getDefault(newKeys(seriesKeys, seriesOrder)),
+        ];
       },
-      getProps: transformedSeries => ({
-        items: transformedSeries.map(s => s.card),
```

Comment:
Oh just realize we don't use `items` in the Funnel chart. I couldn't find a trace of where this has been used. But good that you removed unused props 👍